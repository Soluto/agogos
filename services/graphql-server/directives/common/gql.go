package common

import (
	"agogos/directives/middlewares"
	"agogos/server"
	"context"

	"agogos/utils"

	"github.com/graphql-go/graphql"
	gqlclient "github.com/machinebox/graphql"
	"github.com/sirupsen/logrus"
	"github.com/vektah/gqlparser/ast"
)

type gqlParams struct {
	url       string
	queryName string
}

var gqlMiddleware = middlewares.DirectiveDefinition{
	MiddlewareFactory: func(s server.ServerContext, f *ast.FieldDefinition, d *ast.Directive) middlewares.Middleware {
		params := parseGqlParams(d)
		client := createGqlClient(params.url)

		return middlewares.ConcurrentLeaf(func(rp graphql.ResolveParams) (interface{}, error) {
			request, err := createGqlRequest(s, params, rp)
			if err != nil {
				return nil, err
			}

			return sendRequest(request, client, params)
		})
	},
}

func parseGqlParams(d *ast.Directive) gqlParams {
	params := gqlParams{}
	var ok bool

	arguments := d.ArgumentMap(make(map[string]interface{}))

	params.url, ok = arguments["url"].(string)
	if !ok {
		logrus.Panic("url argument is missing from gql directive")
	}

	params.queryName, ok = arguments["queryName"].(string)
	if !ok {
		logrus.Panic("queryName argument is missing from gql directive")
	}

	return params
}

func createGqlClient(url string) *gqlclient.Client {
	client := gqlclient.NewClient(url)
	return client
}

func createGqlRequest(s server.ServerContext, gqlParams gqlParams, rp graphql.ResolveParams) (*gqlclient.Request, error) {
	query := utils.ResolveParamsToSDLQuery(gqlParams.queryName, rp)
	request := gqlclient.NewRequest(query)
	return request, nil
}

func sendRequest(request *gqlclient.Request, client *gqlclient.Client, gqlParams gqlParams) (interface{}, error) {
	ctx := context.Background()
	var respData interface{}
	if err := client.Run(ctx, request, &respData); err != nil {
		logrus.WithError(err).Panic("error while graphql request")
	}
	result := respData.(map[string]interface{})[gqlParams.queryName]
	return result, nil
}
