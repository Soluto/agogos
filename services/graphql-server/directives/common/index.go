package common

import "agogos/directives/middlewares"

// MiddlewareDefinitions common
var MiddlewareDefinitions = getMiddlewareDefinitions()

func getMiddlewareDefinitions() map[string]middlewares.MiddlewareDefinition {
	directives := make(map[string]middlewares.MiddlewareDefinition)
	directives["log"] = log
	directives["overrideContext"] = overrideContext
	directives["stub"] = stub
	directives["http"] = httpMiddleware
	directives["select"] = selectMiddleware
	directives["alias"] = aliasMiddleware
	return directives
}
