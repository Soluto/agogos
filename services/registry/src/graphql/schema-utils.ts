import {
    DefinitionNode,
    DocumentNode,
    FieldDefinitionNode,
    InputValueDefinitionNode
} from "graphql";
import {
    clone,
    equals,
    flatten,
    groupBy,
    indexBy,
    map,
    pipe,
    unionWith,
    uniq,
    values
} from "ramda";

function validateSameFieldDefs(
    fieldDefA: FieldDefinitionNode | InputValueDefinitionNode,
    fieldDefB: FieldDefinitionNode | InputValueDefinitionNode
) {
    const { directives: _, ...fieldDefWithNoDirectivesA } = fieldDefA;
    const { directives: __, ...fieldDefWithNoDirectivesB } = fieldDefB;
    if (!equals(fieldDefWithNoDirectivesA, fieldDefWithNoDirectivesB)) {
        throw new Error(`Same field definition should be identical except for directives - field ${
            fieldDefA.name.value
            }`);
    }
}

function mergeSameFieldDefinitions(
    fieldDefinitions: ReadonlyArray<
        FieldDefinitionNode | InputValueDefinitionNode
    >
): ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode> {
    if (fieldDefinitions.length === 0) { return []; }
    if (fieldDefinitions.length === 1) { return fieldDefinitions; }

    const [fieldDefA, fieldDefB, ...rest] = fieldDefinitions;
    validateSameFieldDefs(fieldDefA, fieldDefB);

    const { directives: _, ...otherFields } = fieldDefA;
    const fieldWithMergedDirectives = {
        ...otherFields,
        directives: uniq([
            ...(fieldDefA.directives || []),
            ...(fieldDefB.directives || [])
        ])
    };

    return mergeSameFieldDefinitions([fieldWithMergedDirectives, ...rest]);
}

const mergeFields = (
    l: ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode>,
    r: ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode>
) =>
    pipe(
        groupBy(
            (fieldDef: FieldDefinitionNode | InputValueDefinitionNode) =>
                fieldDef.name.value
        ),
        values,
        map(mergeSameFieldDefinitions),
        flatten
    )([...l, ...r]);

export function mergeDocuments(schemas: DocumentNode[]) {
    return clone(schemas).reduce((acc: DocumentNode, next: DocumentNode) => {
        const index = indexBy((x: DefinitionNode) => {
            if (
                x.kind === "ObjectTypeDefinition" ||
                x.kind === "UnionTypeDefinition" ||
                x.kind === "InputObjectTypeDefinition"
            ) {
                return `${x.kind}_${x.name.value}`;
            }
            return "";
        })(acc.definitions);

        const get = (kind: string, value: string) => index[`${kind}_${value}`];

        next.definitions.forEach(x => {
            if (
                (x.kind !== "ObjectTypeDefinition" &&
                    x.kind !== "UnionTypeDefinition" &&
                    x.kind !== "InputObjectTypeDefinition") ||
                !get(x.kind, x.name.value)
            ) {
                (acc.definitions as any).push(x);
                return;
            }

            const ref = get(x.kind, x.name.value);
            if (
                x.kind === "ObjectTypeDefinition" ||
                x.kind === "InputObjectTypeDefinition"
            ) {
                (ref as any).directives = uniq([
                    ...(ref as any).directives,
                    ...(x.directives || [])
                ]);
                (ref as any).fields = mergeFields((ref as any).fields, x.fields || []);

                if (x.kind === "ObjectTypeDefinition") {
                    (ref as any).interfaces = uniq([
                        ...(ref as any).interfaces,
                        ...(x.interfaces || [])
                    ]);
                }
                return;
            }

            if (x.kind === "UnionTypeDefinition") {
                (ref as any).types = unionWith(equals, (ref as any).types, x.types || []);
                return;
            }
        });

        return acc;
    });
}
