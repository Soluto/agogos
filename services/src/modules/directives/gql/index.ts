import { sdl, GqlDirective } from './directive';
import { updateRemoteGqlSchemas, RemoteSchema } from './introspection';

export { sdl, GqlDirective, updateRemoteGqlSchemas, RemoteSchema };
