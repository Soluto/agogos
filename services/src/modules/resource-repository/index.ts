export * from './types';

export { ResourceRepository } from './resource-repository';
export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3-repository';
export { FileSystemResourceRepository } from './fs-repository';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates, applyResourceGroupDeletions } from './actions';
export { default as getResourceRepository } from './get-resource-repository';
