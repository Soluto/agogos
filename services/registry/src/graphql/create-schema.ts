import gql from "graphql-tag";
import { mergeDocuments } from "./schema-utils";
import { DocumentNode } from "graphql";

type GqlSource = {
  source: string;
  gql: string;
};

export type GqlSources = { [source: string]: string };

function parseGqlSource(gqlSource: GqlSource) {
  try {
    return gql(gqlSource.gql);
  } catch (e) {
    throw {
      message: `Failed to load schema for source -  ${gqlSource.source} - ${
        e.message
      }`,
      schema: gqlSource,
      error: e
    };
  }
}

function mergeAllDocuments(docs: DocumentNode[]) {
  try {
    return mergeDocuments(docs);
  } catch (e) {
    throw {
      message: `Failed to merge schemas - ${e.message}`,
      schemas: docs,
      error: e
    };
  }
}

export const makeGqlDocumentFromGqlSources = (gqlSources: GqlSources) => {
  const documentNodes = Object.entries(gqlSources).map(([source, gql]) =>
    parseGqlSource({ source, gql })
  );
  return mergeAllDocuments(documentNodes);
};
