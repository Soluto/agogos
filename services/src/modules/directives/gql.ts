import { SchemaDirectiveVisitor, makeRemoteExecutableSchema, delegateToSchema, introspectSchema } from 'graphql-tools';
import { GraphQLField, GraphQLSchema } from 'graphql';
import { gql } from 'apollo-server-core';
import { ApolloLink } from 'apollo-link';
import { HttpLink } from 'apollo-link-http';
import { RetryLink } from 'apollo-link-retry';
import { setContext } from 'apollo-link-context';
import fetch from 'node-fetch';
import { FastifyRequest } from 'fastify';
import { inject } from '../arguments-injection';
import { ActiveDirectoryAuth, getAuthHeaders } from '../upstreams/authentication';
import logger from '../logger';
import { RequestContext } from '../context';
import { ResourceGroup } from '../resource-repository';

const pendingIntrospectionRetries = new Map<string, NodeJS.Timeout>();

export class GqlDirective extends SchemaDirectiveVisitor {
  createRemoteSchema(
    url: string,
    resourceGroup: ResourceGroup,
    activeDirectoryAuth: ActiveDirectoryAuth,
    timeoutMs?: number
  ) {
    const httpLink: ApolloLink = new HttpLink({
      uri: url,
      fetch: fetch as any,
      fetchOptions: { timeout: timeoutMs ?? 10000 },
    });
    const authLink = setContext(async (_, context) => ({
      headers: await getAuthHeaders(
        resourceGroup,
        activeDirectoryAuth,
        new URL(url).host,
        context?.graphqlContext?.request as FastifyRequest
      ),
    })).concat(httpLink);
    const retryLink = new RetryLink({
      delay: { max: 5000 },
      attempts: {
        max: 5,
        retryIf(error) {
          logger.warn({ error, url }, 'Failed fetching introspection query');
          return !!error;
        },
      },
    }).concat(authLink);

    pendingIntrospectionRetries.delete(url);

    const result: RemoteSchemaWithStatus = { ready: false };
    this.runIntrospection(retryLink, authLink, result, url);
    return result;
  }

  runIntrospection(retryLink: ApolloLink, authLink: ApolloLink, result: RemoteSchemaWithStatus, url: string) {
    // Only introspection should retry, because if we don't have the introspection result this entire resolver will fail.
    // Normal gql requests should not retry, at least for now, because that is more complicated than introspection.
    introspectSchema(retryLink)
      .then(schema => makeRemoteExecutableSchema({ schema, link: authLink }))
      .then(schema => Object.assign(result, { schema, ready: true }))
      .catch(err => {
        logger.error({ err, url }, 'Failed all retries of fetching introspection query, will retry in 10 minutes');
        // https://github.com/Microsoft/TypeScript/issues/30128
        const timeout = global.setTimeout(() => {
          this.runIntrospection(retryLink, authLink, result, url);
        }, 10 * 60 * 1000);
        pendingIntrospectionRetries.set(url, timeout);
      });
  }

  visitFieldDefinition(field: GraphQLField<unknown, RequestContext>) {
    const { url, fieldName, arguments: gqlArgs, operationType: operationTypeParam, timeoutMs } = this.args;
    const operationType = operationTypeParam?.toLowerCase() ?? 'query';
    const remoteSchema = this.createRemoteSchema(
      url,
      this.context.resourceGroup,
      this.context.activeDirectoryAuth,
      timeoutMs
    );

    field.resolve = async (source, args, context, info) => {
      if (!remoteSchema.ready) throw new Error(`Failed reaching remote gql server for introspection (url ${url})`);

      return await delegateToSchema({
        schema: remoteSchema.schema!,
        operation: operationType,
        fieldName,
        args: inject(gqlArgs, { source, args, context, info }) as Record<string, unknown>,
        context,
        info,
      });
    };
  }
}

type RemoteSchemaWithStatus = {
  ready: boolean;
  schema?: GraphQLSchema;
};

export const sdl = gql`
  enum GraphQLOperationType {
    Query
    Mutation
  }

  directive @gql(
    url: String!
    fieldName: String!
    operationType: GraphQLOperationType
    arguments: JSONObject
    timeoutMs: Int
  ) on FIELD_DEFINITION
`;
