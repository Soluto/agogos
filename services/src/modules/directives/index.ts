import { concatAST } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import { sdl as stubSdl, StubDirective } from './stub';
import { sdl as localResolverSdl, LocalResolverDirective } from './local-resolver';
import { sdl as restSdl, RestDirective } from './rest';
import { sdl as gqlSdl, GqlDirective } from './gql';
import { sdl as exportSdl, ExportDirective } from './export';
import { sdl as selectSdl, SelectDirective } from './select';
import {
  policyBaseSdl,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  PolicyDirective,
  PoliciesDirective,
  PolicyQueryDirective,
  policyScalarResolvers,
} from './policy';

export const directiveMap: { [visitorName: string]: typeof SchemaDirectiveVisitor } = {
  stub: StubDirective,
  rest: RestDirective,
  gql: GqlDirective,
  export: ExportDirective,
  select: SelectDirective,
  policy: PolicyDirective,
  policies: PoliciesDirective,
  policyQuery: PolicyQueryDirective,
  localResolver: LocalResolverDirective,
};

export const sdl = concatAST([
  stubSdl,
  restSdl,
  gqlSdl,
  exportSdl,
  selectSdl,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  localResolverSdl,
]);

export { policyScalarResolvers, policyBaseSdl };
