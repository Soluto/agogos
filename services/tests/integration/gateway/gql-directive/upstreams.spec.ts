import { createTestClient, ApolloServerTestClient } from 'apollo-server-testing';
import { graphqlSync, print, printSchema } from 'graphql';
import { gql } from 'apollo-server-core';
import * as nock from 'nock';
import { makeExecutableSchema } from 'graphql-tools';
import createStitchGateway from '../../../../src/modules/apollo-server';
import { ResourceGroup, Schema, Upstream } from '../../../../src/modules/resource-repository';
import { beforeEachDispose } from '../../before-each-dispose';
import { RemoteSchema } from '../../../../src/modules/directives/gql';

jest.mock('../../../../src/modules/resource-repository/get-resource-repository');
import getResourceRepository from '../../../../src/modules/resource-repository/get-resource-repository';

interface TestCase {
  upstreams: Upstream[];
  virtualHost?: string;
  mockAuth?: boolean;
}

const host = 'http://virtual-host';
const defaultRemoteServer = 'http://remote-server';

const testCases: [string, TestCase][] = [
  ['No upstreams', { upstreams: [] }],
  [
    'Upstream with host without targetOrigin',
    {
      upstreams: [
        {
          metadata: { namespace: 'upstreams', name: 'upsteam-1' },
          host: new URL(defaultRemoteServer).host,
        },
      ],
    },
  ],
  [
    'Upstream with host and targetOrigin',
    {
      upstreams: [
        {
          metadata: { namespace: 'upstreams', name: 'upsteam-2' },
          host: new URL(host).host,
          targetOrigin: defaultRemoteServer,
        },
      ],
      virtualHost: 'http://virtual-host',
    },
  ],
];

describe.each(testCases)('Gql - Upstreams', (testCaseName, { upstreams, virtualHost }) => {
  let client: ApolloServerTestClient;
  const remoteServer = virtualHost ?? defaultRemoteServer;

  beforeEachDispose(async () => {
    const remoteSchema = makeExecutableSchema({
      typeDefs: gql`
        type Query {
          foo: String
        }
      `,
      resolvers: {
        Query: {
          foo: () => 'FOO',
        },
      },
    });

    nock(defaultRemoteServer)
      .post('/graphql')
      .reply(200, (_, body: any) =>
        graphqlSync({
          schema: remoteSchema,
          source: body.query,
          variableValues: body.variables,
          operationName: body.operationName,
        })
      );

    const schema: Schema = {
      metadata: {
        namespace: 'upstreams',
        name: 'simple',
      },
      schema: print(gql`
        type Query {
          foo: String @gql(url: "${remoteServer}/graphql", fieldName: "foo")
        }
      `),
    };

    const remoteSchemaResource: RemoteSchema = {
      url: `${remoteServer}/graphql`,
      schema: printSchema(remoteSchema),
    };

    const resourceGroup: ResourceGroup = {
      schemas: [schema],
      upstreams,
      upstreamClientCredentials: [],
      policies: [],
      remoteSchemas: [remoteSchemaResource],
    };

    (getResourceRepository as jest.Mock).mockImplementationOnce(
      jest.fn().mockReturnValueOnce({
        fetchLatest: () => Promise.resolve({ resourceGroup, isNew: true }),
      })
    );

    const { server } = await createStitchGateway();
    client = createTestClient(server);

    return () => {
      nock.cleanAll();
      return server.stop();
    };
  });

  it(testCaseName, async () => {
    const response = await client
      .query({
        query: gql`
          query {
            foo
          }
        `,
      })
      .catch(e => e.response);

    expect(response).toMatchSnapshot();
  });
});
