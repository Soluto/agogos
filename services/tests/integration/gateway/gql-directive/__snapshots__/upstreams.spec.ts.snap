// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gql - Upstreams No upstreams 1`] = `
Object {
  "data": Object {
    "foo": "FOO",
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Gql - Upstreams Upstream with host and targetOrigin 1`] = `
Object {
  "data": Object {
    "foo": "FOO",
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Gql - Upstreams Upstream with host without targetOrigin 1`] = `
Object {
  "data": Object {
    "foo": "FOO",
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
